protoc-gen-kotlin/lib/src/jvmTest/kotlin/CodeGeneratorTest.kt

```diff
@@ -0,0 +1,123 @@
+
+        val protoc: String = (System.getProperty("protoc.path")?.let { "$it/bin/protoc" } ?: "protoc")
+            .let { protoc ->
+                try {
+                    // try executing the protoc binary, this code throws an exception if it cannot be found
+                    ProcessBuilder(protoc).start()
+                    protoc
+                } catch (e: Exception) {
+                    downloadedProtoc()
+                }
+            }
+
+        private fun downloadedProtoc() = try {
+            val filename = "protoc-$DOWNLOAD_PROTOC_VERSION-${getOS()}-${getArch()}.exe"
+            val protocTemp = File.createTempFile("protoc", "tmp")
+            val protoUrl =
+                URL("https://repo1.maven.org/maven2/com/google/protobuf/protoc/$DOWNLOAD_PROTOC_VERSION/$filename")
+            protocTemp.writeBytes(protoUrl.readBytes())
+            protocTemp.setExecutable(true)
+            protocTemp.absolutePath
+        } catch (e: Exception) {
+            throw Exception("failed to download protoc version $DOWNLOAD_PROTOC_VERSION", e)
+        }
+
+        private fun getOS() = System.getProperty("os.name").toLowerCase().let {
+            when {
+                it.contains("windows") -> "windows"
+                it.contains("mac os x") || it.contains("darwin") || it.contains("osx") -> "osx"
+                else -> "linux"
+            }
+        }
+
+        private fun getArch() = System.getProperty("os.arch").toLowerCase().let {
+            when {
+                it.contains("amd64") || it.contains("x86_64") -> "x86_64"
+                else -> "x86_32"
+            }
+        }
+    }
+
+    private fun getFileDescriptorProto(protoFile: String): List<FileDescriptorProto> {
+        val plugin = File(PROTOC_PLUGIN_PATH + ".bat".takeIf {
+            getOS() == "windows"
+        }.orEmpty())
+
+        val proc = ProcessBuilder(
+            protoc,
+            "--proto_path=$PROTO_PATH",
+            "--plugin=protoc-gen-kotlin=${plugin.absoluteFile}",
+            "--descriptor_set_out=$DESCRIPTOR_SET_OUT",
+            protoFile
+        )
+            .redirectOutput(Redirect.PIPE)
+            .redirectError(Redirect.PIPE)
+            .start()
+            .also { it.waitFor(10, TimeUnit.SECONDS) }
+
+        if (proc.exitValue() != 0) {
+            throw IllegalArgumentException("failed generating proto descriptor set for '$protoFile'")
+        }
+
+        return FileDescriptorSet.decodeFromByteArray(File(DESCRIPTOR_SET_OUT).inputStream().readAllBytes()).file
+    }
+
+    private fun compileProto(inputProto: String): KotlinCompilation.Result {
+        val gen = runGenerator(
+            CodeGeneratorRequest(
+                fileToGenerate = listOf(inputProto),
+                protoFile = getFileDescriptorProto(inputProto)
+            )
+        )
+
+        val kotlinSource = SourceFile.kotlin("test.kt", gen.file.first().content!!)
```

Comment:
You should use `gen.file.first().name` rather than `test.kt` as the filename

Reply:
I had to adapt it still since the compiler does not like directories in the path `foobar/baz.kt` (only `baz.kt` will work)